CA1862: Prefer Dictionary<K, V>.TryAddValue(key) over guarded Add(key)

CollinAlpert opened the pull request:
<!--

Make sure you have read the contribution guidelines: 
- https://docs.microsoft.com/contribute/dotnet/dotnet-contribute-code-analysis#contribute-docs-for-caxxxx-rules
- https://github.com/dotnet/roslyn-analyzers/blob/main/GuidelinesForNewRules.md

If your Pull Request is doing one of the following:

- Adding a new diagnostic analyzer or a code fix
- Adding or updating resource strings used by analyzers and code fixes
- Updating analyzer package versions in [Versions.props](../eng/Versions.props)

Then, make sure to run `msbuild /t:pack /v:m` in the repository root; otherwise, the CI build will fail.

- This command must be run from a Visual Studio Developer Command Prompt
- Alternatively, `dotnet msbuild RoslynAnalyzers.sln -t:pack -v:m` can be used from a standard terminal window

Note: Consider merging the PR base branch (`2.9.x`, `main`, or `release/*`) into your branch before you run the pack command to reduce merge conflicts.

-->
This PR adds an Analyzer to detect usages of `dictionary.Add()` guarded by a `dictionary.ContainsKey()` check. It also adds a CodeFix which offers to refactor this construct into a single `dictionary.TryAdd()` call.

Fixes dotnet/runtime#33799, https://github.com/dotnet/roslyn-analyzers/issues/6589

CollinAlpert commented:
Ping for review :)

Therzok commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddTests.cs:
```diff
@@ -0,0 +1,355 @@
+
+namespace Microsoft.NetCore.Analyzers.Performance.UnitTests
+{
+    public class PreferDictionaryTryAddValueOverGuardedAddTests
```
Would something like be flagged as well? 

```csharp
if (!dictionary.TryGetValue(key, out var value))
{
    dictionary.Add(key, new object());
}
// the version below allocations even if the value is already in the map
dictionary.TryAdd(key, new object());
```

In some cases, the hashing might be preferred to allocating indiscriminately.

CollinAlpert replied:
First of all, this analyzer flags usages of `ContainsKey`, not `TryGetValue`.

To answer your question, no, the analyzer would not flag this construct. It searches for literals only ([see here](https://github.com/CollinAlpert/roslyn-analyzers/blob/392174cdf469ef76d822e8308d8b0c75717c942b/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs#L97)), as per this requirement: https://github.com/dotnet/runtime/issues/33799#issuecomment-773499679

Therzok replied:
Thank you!

Miizukii replied:
looking at that comment: 
> [dotnet/runtime#33799 (comment)](https://github.com/dotnet/runtime/issues/33799#issuecomment-773499679)
where the associated value is: (a) a literal; or (b) already fully hydrated before the call to ContainsKey

would be great if you could include (b) as only literals seems quite limited

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+            RuleId,
+            s_localizableTitle,
+            s_localizableTryGetValueMessage,
+            "Performance",
```
Use https://github.com/dotnet/roslyn-analyzers/blob/adb9c1cf0416e1dc2b547b7a962a171e54cfb85d/src/Utilities/Compiler/DiagnosticCategory.cs#L11

CollinAlpert replied:
Neat!

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+        private const string ContainsKey = nameof(IDictionary<dynamic, dynamic>.ContainsKey);
+        private const string Add = nameof(IDictionary<dynamic, dynamic>.Add);
+
+        private static readonly LocalizableString s_localizableTitle = CreateLocalizableResourceString(nameof(PreferDictionaryTryAddValueTitle));
+        private static readonly LocalizableString s_localizableTryGetValueMessage = CreateLocalizableResourceString(nameof(PreferDictionaryTryAddValueMessage));
+        private static readonly LocalizableString s_localizableTryGetValueDescription = CreateLocalizableResourceString(nameof(PreferDictionaryTryAddValueDescription));
```
inline

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+            "Performance",
+            RuleLevel.IdeSuggestion,
+            s_localizableTryGetValueDescription,
+            false,
+            false);
```
named argument

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+            [NotNullWhen(true)] out IMethodSymbol? containsKeySymbol,
+            [NotNullWhen(true)] out IMethodSymbol? addSymbol)
+        {
+            containsKeySymbol = null;
+            addSymbol = null;
+            iDictionaryType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2);
+            if (iDictionaryType is null)
+            {
+                return false;
+            }
```
Does it work like that?

```suggestion
            iDictionaryType = WellKnownTypeProvider.GetOrCreate(compilation).GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemCollectionsGenericIDictionary2);
            if (iDictionaryType is null)
            {
                containsKeySymbol = null;
                addSymbol = null;
                return false;
            }
```

CollinAlpert replied:
Yup

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetDictionaryTypeAndMembers(context.Compilation, out var iDictionaryType, out var containsKeySymbol, out var addSymbol))
+            {
+                return;
+            }
```
I think we should also return if there is no `TryAdd` method.

CollinAlpert replied:
Since `IDictionary` does not have the `TryAdd` method, I don't think this would be the right place to add the check. I could add it in the `IsDictionaryType` method. What do you think?

Youssef1313 replied:
Yes, the idea is that we don't give a warning suggesting to use `TryAdd` where such method doesn't actually exist.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+namespace Microsoft.NetCore.Analyzers.Performance
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class PreferDictionaryTryAddValueOverGuardedAddAnalyzer : DiagnosticAnalyzer
```
Is "Value" in the name extra? Consider removing it, and from resource names as well.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+                   || suspectedDictionaryType.AllInterfaces.Any((@interface, dictionary) => @interface.OriginalDefinition.Equals(dictionary, SymbolEqualityComparer.Default), iDictionaryType);
+        }
+
+        private static bool DoesSignatureMatch(IMethodSymbol suspected, IMethodSymbol comparator)
+        {
+            return suspected.OriginalDefinition.ReturnType.Name == comparator.ReturnType.Name
+                   && suspected.Name == comparator.Name
+                   && suspected.Parameters.Length == comparator.Parameters.Length
+                   && suspected.Parameters.Zip(comparator.Parameters, (p1, p2) => p1.OriginalDefinition.Type.Name == p2.Type.Name).All(isParameterEqual => isParameterEqual);
+        }
```
Why not use `suspected.Equals(comparator, SymbolEqualityComparer.Default)`?

CollinAlpert replied:
Because symbolically, `System.Collections.Generic.Dictionary<TKey, TValue>.ContainsKey(TKey)` and `System.Collections.Generic.IDictionary<TKey, TValue>.ContainsKey(TKey)` are not equal.

Youssef1313 replied:
@CollinAlpert There is an `IsImplementationOfInterfaceMember` extension method that I think could work for this scenario.

CollinAlpert replied:
Yeah, I tried substituting the body of the method with `return suspected.OriginalDefinition.IsImplementationOfInterfaceMember(comparator);`, but that also doesn't work.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+
+        private static void OnOperationAnalysis(INamedTypeSymbol iDictionaryType, IMethodSymbol containsKeySymbol, IMethodSymbol addSymbol, OperationAnalysisContext context)
+        {
+            var conditional = (IConditionalOperation)context.Operation;
+            if (conditional.Condition is IInvocationOperation invocation
+                && IsDictionaryType(invocation.TargetMethod.ContainingType, iDictionaryType)
+                && DoesSignatureMatch(invocation.TargetMethod, containsKeySymbol)
+                && TryGetEligibleDictionaryAddLocation(conditional.WhenFalse, invocation.Instance, invocation.Arguments[0].Value, addSymbol, out var dictionaryAddLocation))
+            {
+                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, invocation.Syntax.GetLocation(), additionalLocations));
+            }
+            else if (conditional.Condition is IUnaryOperation { OperatorKind: UnaryOperatorKind.Not, Operand: IInvocationOperation i }
+                     && IsDictionaryType(i.TargetMethod.ContainingType, iDictionaryType)
+                     && DoesSignatureMatch(i.TargetMethod, containsKeySymbol)
+                     && TryGetEligibleDictionaryAddLocation(conditional.WhenTrue, i.Instance, i.Arguments[0].Value, addSymbol, out dictionaryAddLocation))
+            {
+                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, i.Syntax.GetLocation(), additionalLocations));
+            }
```
A little bit of refactoring to avoid duplication.

```suggestion
            var conditional = (IConditionalOperation)context.Operation;
            var condition = conditional.Condition;
            var whenBranch = conditional.WhenFalse;
            if (condition is IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } unaryOperation)
            {
                condition = unaryOperation.Operand;
                whenBranch = conditional.WhenTrue;
            }

            if (condition is IInvocationOperation invocation
                && IsDictionaryType(invocation.TargetMethod.ContainingType, iDictionaryType)
                && DoesSignatureMatch(invocation.TargetMethod, containsKeySymbol)
                && TryGetEligibleDictionaryAddLocation(conditional.WhenFalse, invocation.Instance, invocation.Arguments[0].Value, addSymbol, out var dictionaryAddLocation))
            {
                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, invocation.Syntax.GetLocation(), additionalLocations));
            }
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+                && TryGetEligibleDictionaryAddLocation(conditional.WhenFalse, invocation.Instance, invocation.Arguments[0].Value, addSymbol, out var dictionaryAddLocation))
+            {
+                var additionalLocations = ImmutableArray.Create(dictionaryAddLocation);
+                context.ReportDiagnostic(Diagnostic.Create(DiagnosticDescriptor, invocation.Syntax.GetLocation(), additionalLocations));
```
Use `invocation.CreateDiagnostic` instead.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+                return false;
+            }
+
+            static bool OperationEquals(SemanticModel semanticModel, IOperation left, IOperation right)
+            {
+                var leftSymbol = semanticModel.GetSymbolInfo(left.Syntax).Symbol;
+                var rightSymbol = semanticModel.GetSymbolInfo(right.Syntax).Symbol;
+
+                return SymbolEqualityComparer.Default.Equals(leftSymbol, rightSymbol);
+            }
+
+            bool IsEligibleDictionaryAddInvocation(IInvocationOperation invocation)
+            {
+                return OperationEquals(invocation.SemanticModel, invocation.Instance, dictionaryInstance)
+                       && DoesSignatureMatch(invocation.TargetMethod, addSymbol)
+                       && invocation.Arguments[0].Value.Syntax.IsEquivalentTo(containsKeyArgument.Syntax)
+                       && invocation.Arguments[1].Value.Kind == OperationKind.Literal;
+            }
```
Move local functions at the end of the method.

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+                return false;
+            }
+
+            containsKeySymbol = iDictionaryType.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(m => m.Name == ContainsKey);
```
```suggestion
            containsKeySymbol = iDictionaryType.GetMembers(ContainsKey).OfType<IMethodSymbol>().FirstOrDefault();
```

Youssef1313 commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+            }
+
+            containsKeySymbol = iDictionaryType.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(m => m.Name == ContainsKey);
+            addSymbol = iDictionaryType.GetMembers().OfType<IMethodSymbol>().FirstOrDefault(m => m.Name == Add);
```
```suggestion
            addSymbol = iDictionaryType.GetMembers(Add).OfType<IMethodSymbol>().FirstOrDefault();
```

Youssef1313 commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferDictionaryTryAddValueOverGuardedAddFixer.cs:
```diff
@@ -0,0 +1,103 @@
+    {
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.FirstOrDefault();
+            var dictionaryAddLocation = diagnostic?.AdditionalLocations[0];
+            if (dictionaryAddLocation is null)
+            {
+                return;
+            }
```
```suggestion
            var diagnostic = context.Diagnostics[0];
            var dictionaryAddLocation = diagnostic.AdditionalLocations[0];
```

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferDictionaryTryAddValueOverGuardedAddFixer.vb:
```diff
@@ -0,0 +1,85 @@
+        Inherits PreferDictionaryTryAddValueOverGuardedAddFixer
+
+        Public Overrides Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
+            Dim diagnostic = context.Diagnostics.FirstOrDefault()
```
```suggestion
            Dim diagnostic = context.Diagnostics(0)
```

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferDictionaryTryAddValueOverGuardedAddFixer.vb:
```diff
@@ -0,0 +1,85 @@
+
+        Public Overrides Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
+            Dim diagnostic = context.Diagnostics.FirstOrDefault()
+            Dim dictionaryAddLocation = diagnostic?.AdditionalLocations(0)
```
```suggestion
            Dim dictionaryAddLocation = diagnostic.AdditionalLocations(0)
```

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferDictionaryTryAddValueOverGuardedAddFixer.vb:
```diff
@@ -0,0 +1,85 @@
+        Public Overrides Async Function RegisterCodeFixesAsync(context As CodeFixContext) As Task
+            Dim diagnostic = context.Diagnostics.FirstOrDefault()
+            Dim dictionaryAddLocation = diagnostic?.AdditionalLocations(0)
+            If dictionaryAddLocation Is Nothing
+                Return
+            End If
```
```suggestion
```

Youssef1313 commented on src/NetAnalyzers/VisualBasic/Microsoft.NetCore.Analyzers/Performance/BasicPreferDictionaryTryAddValueOverGuardedAddFixer.vb:
```diff
@@ -0,0 +1,85 @@
+            Dim containsKeyInvocation = TryCast(root.FindNode(context.Span), InvocationExpressionSyntax)
+            Dim containsKeyAccess = TryCast(containsKeyInvocation?.Expression, MemberAccessExpressionSyntax)
+            If _
+                dictionaryAddInvocation Is Nothing Or
+                containsKeyInvocation Is Nothing Or
```
```suggestion
                dictionaryAddInvocation Is Nothing OrElse
                containsKeyInvocation Is Nothing OrElse
```

Youssef1313 commented on src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddTests.cs:
```diff
@@ -0,0 +1,362 @@
+    }}
+}}";
+
+        private const string CheckNegativeReturnedValue = @"
+if (!{|#0:d.ContainsKey(5)|})
+{
+	{|#1:d.Add(5, 6)|};
+}";
```
Is there a test where `ContainsKey` and `Add` are invoked on two different dictionaries?

CollinAlpert replied:
There is now :)

CollinAlpert commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+
+            static bool OperationEquals(SemanticModel semanticModel, IOperation left, IOperation right)
+            {
+                var leftSymbol = semanticModel.GetSymbolInfo(left.Syntax).Symbol;
+                var rightSymbol = semanticModel.GetSymbolInfo(right.Syntax).Symbol;
+
+                return SymbolEqualityComparer.Default.Equals(leftSymbol, rightSymbol);
```
Cool! I'll put this into an extension method.

CollinAlpert commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAddAnalyzer.cs:
```diff
@@ -0,0 +1,149 @@
+
+            if (conditionalBody is IBlockOperation block)
+            {
+                dictionaryAddLocation = block.Operations.OfType<IExpressionStatementOperation>().FirstOrDefault(e => e.Operation is IInvocationOperation i && IsEligibleDictionaryAddInvocation(i))?.Operation.Syntax.GetLocation();
```
Thanks! Will look into this and add some tests.

Miizukii commented on src/Utilities/Compiler/Extensions/IOperationExtensions.cs:
```diff
@@ -1136,6 +1136,20 @@ public static bool IsAnyCompoundAssignment(this IOperation operation)
+                IParameterReferenceOperation pa => pa.Parameter,
+                IPropertyReferenceOperation p => p.Property,
+                IMemberReferenceOperation m => m.Member,
+                IArrayElementReferenceOperation a => GetSymbolFromReference(a.ArrayReference),
```
this would be true for data[0] and data[1], since it only compares that data is the same symbol.
I refactored my code to handle the indices directly
https://github.com/dotnet/roslyn-analyzers/blob/41da30d77f263148615211dd61945c94a0da5487/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferDictionaryTryGetValueAnalyzer.cs#L384-L394

CollinAlpert replied:
Can you explain why this would only be true for `data[0]` and `data[1]`? From what I can tell, it fetches the array reference symbol regardless of index, which is exactly the desired behavior.

Miizukii replied:
``` c#
var data = new int[]{1,2};
var x = data[0];
var y = data[1];
```
If you look at `data[0]`, you will get `IArrayElementReferenceOperation` with `ArrayReference` being a `ILocalReferenceOperation` where the `Local` symbol is `data`, but the symbol itself does not tell you how it was used, data[1] has the same Symbol in its `IArrayElementReferenceOperation.ArrayReference.Local`.

To make sure that the operation not only reference the same array but also the same element, you need to test `IArrayElementReferenceOperation.Indices`.
Which here would be `ILiteralOperation` with ConstantValue 0 and 1 respectively, but could also be another reference operation like `data[data[0]]`.

buyaa-n commented:
@CollinAlpert could you please resolve merge conflicts? And fix test/build failures apply remaining feedbacks (if there is any).

CollinAlpert commented:
That is a good point. It will take me a bit longer, but I think it is a good idea. Should the analyzer still emit different diagnostic IDs?

buyaa-n commented:
Yes, diagnostic ids should be different so that users could handle (suppress) them separately and also will be useful for the code fix

CollinAlpert commented:
I resolved the merge conflicts and the remaining comments. If you want to complete this PR first, I could understand. Nonetheless I will start working on the consolidation of the analyzers now.

CollinAlpert commented:
@buyaa-n I have merged the analyzers and would appreciate a final review! :)

buyaa-n commented:
> @buyaa-n I have merged the analyzers and would appreciate a final review! :)

Thank you @CollinAlpert, I did tested with runtime and somehow the fixer is not producing suggestions in VS. Have you tried the use the fixer in VS? Could you take a look into this issue?

buyaa-n commented on nuget/NetAnalyzers/Microsoft.CodeAnalysis.NetAnalyzers.Package.csproj:
```diff
@@ -25,7 +25,7 @@
     <AnalyzerNupkgAssembly Include="Microsoft.CodeAnalysis.CSharp.NetAnalyzers.dll" />
     <AnalyzerNupkgAssembly Include="Microsoft.CodeAnalysis.VisualBasic.NetAnalyzers.dll" />
-    <PackageReadmeFileSource Include="..\..\src\NetAnalyzers\readme.md"/>
+    <PackageReadmeFileSource Include="..\..\src\NetAnalyzers\readme.md" />
```
Please revert unrelated changes

CollinAlpert replied:
Done.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryAddValueOverGuardedAdd.Fixer.cs:
```diff
@@ -0,0 +1,21 @@
+            return WellKnownFixAllProviders.BatchFixer;
+        }
+
+        public override ImmutableArray<string> FixableDiagnosticIds { get; } = ImmutableArray.Create(PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.PreferTryAddRuleId);
```
NIT: The fixers could have been integrated too, have you tried it? As you can see the `FixableDiagnosticIds` is an array, meaning it could support diagnostics with different ids.

CollinAlpert replied:
Done.

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -2054,4 +2054,16 @@ Widening and user defined conversions are not supported with generic types.</val
+    <value>Prefer a 'TryAdd' call over an 'Add' call guarded by a 'ContainsKey' check to avoid double lookup</value>
+  </data>
+  <data name="PreferDictionaryTryAddDescription" xml:space="preserve">
+    <value>Prefer a 'TryAdd' call over an 'Add' call guarded by a 'ContainsKey' check. 'TryAdd' behaves the same as 'Add' except that it doesn't throw an exception when a value for a given key already exists, but rather returns 'false'.</value>
```
Tag @gewarren for messaging

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs:
```diff
@@ -177,74 +265,92 @@ private static bool GetParentConditionalOperation(IOperation operation, ref Dict
 
                     default:
                         conditionalOperation = null;
+
```
NIT: There is lots of extra line added in switch cases, here an below

CollinAlpert replied:
I believe that is due to `IDE2003`. But I removed the blank lines where possible.

gewarren commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -2054,4 +2054,16 @@ Widening and user defined conversions are not supported with generic types.</val
+    <value>Prefer the 'IDictionary.TryAdd(TKey, TValue)' method</value>
+  </data>
+  <data name="PreferDictionaryTryAddMessage" xml:space="preserve">
+    <value>Prefer a 'TryAdd' call over an 'Add' call guarded by a 'ContainsKey' check to avoid double lookup</value>
```
```suggestion
    <value>To avoid double lookup, call 'TryAdd' instead of calling 'Add' with a 'ContainsKey' guard</value>
```

gewarren commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/MicrosoftNetCoreAnalyzersResources.resx:
```diff
@@ -2054,4 +2054,16 @@ Widening and user defined conversions are not supported with generic types.</val
+    <value>Prefer a 'TryAdd' call over an 'Add' call guarded by a 'ContainsKey' check to avoid double lookup</value>
+  </data>
+  <data name="PreferDictionaryTryAddDescription" xml:space="preserve">
+    <value>Prefer a 'TryAdd' call over an 'Add' call guarded by a 'ContainsKey' check. 'TryAdd' behaves the same as 'Add' except that it doesn't throw an exception when a value for a given key already exists, but rather returns 'false'.</value>
```
```suggestion
    <value>Prefer a 'TryAdd' call over an 'Add' call guarded by a 'ContainsKey' check. 'TryAdd' behaves the same as 'Add', except that when the specified key already exists, it returns 'false' instead of throwing an exception.</value>
```

CollinAlpert replied:
Done.

CollinAlpert commented:
> Thank you @CollinAlpert, I did tested with runtime and somehow the fixer is not producing suggestions in VS. Have you tried the use the fixer in VS? Could you take a look into this issue?

That's weird. I double checked everything, and the IDs registered are all correct and passed to the appropriate properties. Unfortunately I don't use VS, so I can't check there. I consolidated the fixers now, maybe that will help?

buyaa-n commented:
Thanks for the updates @CollinAlpert, now the fixer is working in VS, so I continued with analyzing all diagnostics and found these false positives:
1. The analyzers flagging an instance of `ImmutableDictionary<TKey, TValue>.Builder` and suggesting the code fix when it doesn't have a `TryAdd(TKey, TValue)` method, this case should not be flagged
2. The analyzer flagging and suggesting a fixer when the value is only created within the conditional, this should exclude this scenatio. Example code snippet:
```cs
        XmlQualifiedName notationName = GetNameQualified(false);
        SchemaNotation? notation = null;
        if (!_schemaInfo.Notations.ContainsKey(notationName.Name))
        {
            _undeclaredNotations?.Remove(notationName.Name);
            notation = new SchemaNotation(notationName);
            _schemaInfo.Notations.Add(notation.Name.Name, notation);
        }
```
`SchemaNotation notation` only created with this is not exist in the dictionary

Alos found one false negative case, it is not flagging the below case when it should:
```cs
        foreach (KeyValuePair<XmlQualifiedName, SchemaElementDecl> entry in sinfo._elementDecls)
        {
            if (!_elementDecls.ContainsKey(entry.Key))
            {
                _elementDecls.Add(entry.Key, entry.Value);
            }
        }
```
Please fix these issues, thank you!

CollinAlpert commented:
@buyaa-n Thanks for the hints. I have addressed them and added tests.

Regarding the false negative, I don't think it is one. Behind a property reference (in this case `entry.Value`) could be a computation which the developer only wants to trigger when the key is not present in the dictionary. Currently the analyzer only flags literals, local references, field references and constants.

buyaa-n commented:
> @buyaa-n Thanks for the hints. I have addressed them and added tests.

Thank you!

> Regarding the false negative, I don't think it is one. Behind a property reference (in this case entry.Value) could be a computation which the developer only wants to trigger when the key is not present in the dictionary. Currently the analyzer only flags literals, local references, field references and constants.

Well, the `KeyValuePair<TKey,TValue >.Value` just [returns an internal field](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/KeyValuePair.cs#L43) which set within constructor. 
```cs
        public KeyValuePair(TKey key, TValue value)
        {
            this.key = key;
            this.value = value;
        }
        public TKey Key => key;
        public TValue Value => value;
```
Anyway, I guess you meant about any property not specific to `KeyValuePair`, so this scenario is not a blocker for the PR. We can discuss this scenario with a separate issue.

CollinAlpert commented:
Yes, I meant properties in general could have expensive computations in their getter, not specifically KeyValuePair.

buyaa-n commented on src/NetAnalyzers/CSharp/Microsoft.NetCore.Analyzers/Performance/CSharpPreferDictionaryTryMethodsOverContainsKeyGuardFixer.cs:
```diff
@@ -46,12 +62,12 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
                 switch (node)
                 {
-                    case ElementAccessExpressionSyntax:
+                    case ElementAccessExpressionSyntax when diagnostic.Id == PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.PreferTryGetValueRuleId:
```
Seems this method only called when `diagnostic.Id == PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.PreferTryGetValueRuleId` is true, why do we need to add this condition here again?
```suggestion
                    case ElementAccessExpressionSyntax :
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs:
```diff
@@ -96,71 +105,176 @@ public ImmutableArray<Location>.Builder UsageLocations
-                return;
+            void ReportGuardedDictionaryPattern(SearchContext searchContext, DiagnosticDescriptor diagnosticDescriptor)
+            {
+                if (searchContext == SearchContext.AddMethod && suspectedDictionaryType.GetMembers(TryAdd).IsEmpty)
+                {
+                    return;
+                }
```
Does this check need to be here? Could we move this into `else if (searchContext == SearchContext.AddMethod)` conditional block below?

CollinAlpert replied:
I wanted to exit as early as possible if the dictionary does not have a `TryAdd` method. But I can still move it, if you would like me to.

buyaa-n replied:
Ok, you can keep it

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs:
```diff
@@ -96,71 +105,176 @@ public ImmutableArray<Location>.Builder UsageLocations
+                    guardedPath = guardsTruePath ? conditionalOperation.WhenTrue : conditionalOperation.WhenFalse;
+                }
+                else if (searchContext == SearchContext.AddMethod)
+                {
+                    guardedPath = guardsTruePath ? conditionalOperation.WhenFalse : conditionalOperation.WhenTrue;
+                }
```
i.e:
```suggestion
                {
                    if (suspectedDictionaryType.GetMembers(TryAdd).IsEmpty)
                    {
                        return;
                    }
                    guardedPath = guardsTruePath ? conditionalOperation.WhenFalse : conditionalOperation.WhenTrue;
                }
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs:
```diff
@@ -96,71 +105,176 @@ public ImmutableArray<Location>.Builder UsageLocations
         //only handles simple conditions: .. && x.ContainsKey(y) or !x.ContainsKey(y) || ..
-        private static bool GetParentConditionalOperation(IOperation operation, ref DictionaryUsageContext context, [NotNullWhen(true)] out IConditionalOperation? conditionalOperation, out bool guardsTruePath)
+        private static bool GetParentConditionalOperation(IOperation operation, ref DictionaryUsageContext usageContext, SearchContext searchContext,
+            [NotNullWhen(true)]
+            out IConditionalOperation? conditionalOperation, out bool guardsTruePath)
```
NIT: extra line
```suggestion
            [NotNullWhen(true)] out IConditionalOperation? conditionalOperation, out bool guardsTruePath)
```

buyaa-n commented on src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/PreferDictionaryTryMethodsOverContainsKeyGuardAnalyzer.cs:
```diff
@@ -96,71 +105,176 @@ public ImmutableArray<Location>.Builder UsageLocations
+        }
+
+        private static bool TryGetDictionaryTypeAndMembers(
+            Compilation compilation,
+            [NotNullWhen(true)]
+            out INamedTypeSymbol? iDictionaryType,
+            [NotNullWhen(true)]
+            out IMethodSymbol? containsKeySymbol,
+            [NotNullWhen(true)]
+            out IMethodSymbol? addSymbol)
```
```suggestion
            Compilation compilation,
            [NotNullWhen(true)] out INamedTypeSymbol? iDictionaryType,
            [NotNullWhen(true)] out IMethodSymbol? containsKeySymbol,
            [NotNullWhen(true)] out IMethodSymbol? addSymbol)
```

buyaa-n commented on src/NetAnalyzers/Core/AnalyzerReleases.Unshipped.md:
```diff
@@ -15,6 +15,7 @@ CA1859 | Performance | Info | UseConcreteTypeAnalyzer, [Documentation](https://l
 CA1860 | Performance | Info | PreferLengthCountIsEmptyOverAnyAnalyzer, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1860)
 CA1861 | Performance | Info | AvoidConstArrays, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1861)
 CA1862 | Performance | Info | RecommendCaseInsensitiveStringComparison, [Documentation](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1862)
+CA1863 | Performance | Info | PreferDictionaryTryAddValueOverGuardedAddAnalyzer, [Documentation](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1862)
```
NIT:
```suggestion
CA1863 | Performance | Info | PreferDictionaryTryAddValueOverGuardedAddAnalyzer, [Documentation](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1863)
```

buyaa-n reviewed (approved):
Left some NIT, overall LGTM, thank you!

It would be great if you could add a test case for https://github.com/dotnet/roslyn-analyzers/issues/6589
